File: Server/Academia/models/CoursePoll.js
const mongoose = require('mongoose');
const Joi = require('joi');
const { v4: uuidv4 } = require('uuid');

/// CoursePoll Schema
const coursePollSchema = new mongoose.Schema({
    _id: {
        type: String,
        default: uuidv4,
        unique: true,
    },
    courseId: {
        type: String,
        required: true,
        trim: true,
    },
    options : {
        type: [String],
        required: true,
        trim: true,
    },
    votes : {
        type: [Number],
        default: [],
    },
    closePollDate: {
        type: Date,
        required: true,
    },
    totalVoters: {
        type: Number,
        default: 0,
    },
}, {
    _id: false,
    timestamps: {
        currentTime: () => new Date().toISOString(), // Use ISO 8601 format for timestamps
        createdAt: 'createdOn',
        updatedAt: 'editedOn',
    },
    collection: 'CoursePoll',
});

/// CoursePoll Model
const coursePoll = mongoose.model('CoursePoll', coursePollSchema);

/// Validate CoursePoll
function validateCreateCoursePoll(coursePoll) {
    const schema = Joi.object({
        courseId: Joi.string().required(),
        options: Joi.array().items(Joi.string()).required(),
        votes: Joi.array().items(Joi.number()),
        closePollDate: Joi.date().required(),
        totalVoters: Joi.number(),
    });

    return schema.validate(coursePoll);
}

function validateUpdateCoursePoll(coursePoll) {
    const schema = Joi.object({
        courseId: Joi.string(),
        options: Joi.array().items(Joi.string()),
        votes: Joi.array().items(Joi.number()),
        closePollDate: Joi.date(),
        totalVoters: Joi.number(),
    });

    return schema.validate(coursePoll);
}

/// CoursePoll Verbs
const createCoursePollVerb = {
    "id-enum": "create-course-poll",
    display: {
        "en-US": "Created Course Poll"
    }
};

const updateCoursePollVerb = {
    "id-enum": "update-course-poll",
    display: {
        "en-US": "Updated Course Poll"
    }
};

const deleteCoursePollVerb = {
    "id-enum": "delete-course-poll",
    display: {
        "en-US": "Deleted Course Poll"
    }
};

module.exports = {
    coursePoll,
    validateCreateCoursePoll,
    validateUpdateCoursePoll,
    createCoursePollVerb,
    updateCoursePollVerb,
    deleteCoursePollVerb,
};

File: Mobile App/academia/lib/domain/models/course_poll.dart
class CoursePoll {
  String id;
  String courseId;
  List<String> options;
  List<int> votes;
  DateTime closePollDate;
  int totalVoters;

  CoursePoll({
    required this.id,
    required this.courseId,
    required this.options,
    required this.votes,
    required this.closePollDate,
    required this.totalVoters,
  });

  factory CoursePoll.fromJson(Map<String, dynamic> json) {
    return CoursePoll(
      id: json['_id'],
      courseId: json['courseId'],
      options: json['options'].cast<String>(),
      votes: json['votes'].cast<int>(),
      closePollDate: DateTime.parse(json['closePollDate']),
      totalVoters: json['totalVoters'],
    );
  }

  Map<String, dynamic> toJson() {
    return {
      '_id': id,
      'courseId': courseId,
      'options': options,
      'votes': votes,
      'closePollDate': closePollDate.toIso8601String(),
      'totalVoters': totalVoters,
    };
  }
}
